### PR TITLE
implement Capabilities' contains method and new CapabilitiesResolver

### PR DESCRIPTION
**Pull request by mmandlis**
this PR is noop, and mostly just translation of TS code into KT
in the next PR, i'll actually move to using the new resolver and deprecate the old one.

### CODE REVIEW

File: java/arcs/core/storage/CapabilitiesResolverNew.kt [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    factoriesList: List<StorageKeyFactory> = listOf(),
+    val selector: FactorySelector = SimpleCapabilitiesSelector()
+) {
+    val factories: Map<String, StorageKeyFactory>
+
+    init {
+        require(factoriesList.distinctBy { it.protocol }.size == factoriesList.size) {
+            "Storage keys protocol must be unique $factoriesList."
+        }
+        factories = factoriesList.map { it.protocol to it }.toMap().toMutableMap()
```
Comment (csilvestrini) [inline java/arcs/core/storage/CapabilitiesResolverNew.kt:39]: `factoriesList.associateBy { it.protocol }.toMutableMap()`

File: java/arcs/core/storage/CapabilitiesResolverNew.kt [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        CapabilitiesResolverNew.defaultStorageKeyFactories.forEach { (protocol, factory) ->
+            if (factories[protocol] == null) factories.put(protocol, factory)
+        }
+    }
+
+    /* Options used to construct [CapabilitiesResolver] */
```
Comment (csilvestrini) [inline java/arcs/core/storage/CapabilitiesResolverNew.kt:45]: Nit: full stop, and use two asterisks at start: `/**`

File: java/arcs/core/storage/CapabilitiesResolverNew.kt [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        return createStorageKeyWithFactory(requireNotNull(factory), type, handleId)
+    }
+
+    fun createStorageKeyWithFactory(
+        factory: StorageKeyFactory,
+        type: Type,
+        handleId: String
+    ): StorageKey {
+        val containerKey = factory.create(StorageKeyFactory.ContainerStorageKeyOptions(
+            options.arcId, toEntitySchema(type)))
```
Comment (csilvestrini) [inline java/arcs/core/storage/CapabilitiesResolverNew.kt:67]: Nit: line wrapping:

```
val containerKey = factory.create(
    StorageKeyFactory.ContainerStorageKeyOptions(options.arcId, toEntitySchema(type))
)
```

Or if that still doesn't fit:

```
val containerKey = factory.create(
    StorageKeyFactory.ContainerStorageKeyOptions(
        options.arcId,
        toEntitySchema(type)
    )
)
```

File: java/arcs/core/storage/CapabilitiesResolverNew.kt [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        val containerChildKey = containerKey.childKeyForHandle(handleId)
+        if (type is ReferenceType<*>) {
+            return containerChildKey
+        }
+        val backingKey = factory.create(
+            StorageKeyFactory.BackingStorageKeyOptions(options.arcId, toEntitySchema(type)))
```
Comment (csilvestrini) [inline java/arcs/core/storage/CapabilitiesResolverNew.kt:73]: Fix wrapping here too

File: java/arcs/core/storage/CapabilitiesResolverNew.kt [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+            is CollectionType<*> -> if (type.collectionType is EntityType) {
+                return (type.collectionType as EntityType).entitySchema
+            }
+            is ReferenceType<*> -> return type.entitySchema!!
+            is EntityType -> return type.entitySchema
+        }
+        throw IllegalArgumentException("Can't retrieve entitySchema of unknown type $type")
+    }
+
+    // An interface for selecting a factory, if more than one are available for the Capabilities.
```
Comment (csilvestrini) [inline java/arcs/core/storage/CapabilitiesResolverNew.kt:99]: Use KDoc style comment: `/** */`

File: java/arcs/core/storage/CapabilitiesResolverNew.kt [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    interface FactorySelector {
+        fun select(factories: Collection<StorageKeyFactory>): StorageKeyFactory
+    }
+
+    // An implementation of a FactorySelector choosing a factory with a least
+    // restrictive max capabilities set.
```
Comment (csilvestrini) [inline java/arcs/core/storage/CapabilitiesResolverNew.kt:105]: Here too. Then put `FactorySelector` inside square brackets.

File: java/arcs/core/storage/StorageKeyFactory.kt [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+import arcs.core.common.ArcId
+import arcs.core.data.CapabilitiesNew
+import arcs.core.data.Schema
+
+abstract class StorageKeyFactory(val protocol: String, val capabilities: CapabilitiesNew) {
+
+    abstract fun create(options: StorageKeyOptions): StorageKey
+
+    // Returns true, if the current storage key class can support the given set of [CapabilitiesNew].
```
Comment (csilvestrini) [inline java/arcs/core/storage/StorageKeyFactory.kt:22]: Use KDoc style comment: `/** */`

File: java/arcs/core/storage/CapabilitiesResolverNew.kt [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+) {
+    val factories: Map<String, StorageKeyFactory>
+
+    init {
+        require(factoriesList.distinctBy { it.protocol }.size == factoriesList.size) {
+            "Storage keys protocol must be unique $factoriesList."
+        }
+        factories = factoriesList.map { it.protocol to it }.toMap().toMutableMap()
+        CapabilitiesResolverNew.defaultStorageKeyFactories.forEach { protocol, factory ->
+            if (factories[protocol] == null) factories.put(protocol, factory)
```
[the remark this answers is not in the document]
Comment (mmandlis) [reply java/arcs/core/storage/CapabilitiesResolverNew.kt]: this is neat, i might end up liking kotlin after all :) thank you so much for all the comments and advice.

File: java/arcs/core/storage/CapabilitiesResolverNew.kt [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+        }
+    }
+
+    /* Options used to construct [CapabilitiesResolver] */
+    data class Options(val arcId: ArcId)
+
+    fun createStorageKey(capabilities: CapabilitiesNew, type: Type, handleId: String): StorageKey {
+        val selectedFactories =
+            // defaultStorageKeyFactories.filterValues { it.supports(capabilities) } +
```
[the remark this answers is not in the document]
Comment (mmandlis) [reply java/arcs/core/storage/CapabilitiesResolverNew.kt]: sorry, removed.

File: java/arcs/core/storage/CapabilitiesResolverNew.kt [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+        // ReferenceModeStorageKeys in different drivers can cause problems with garbage collection.
+        require(backingKey.protocol == containerKey.protocol) {
+            "Backing and containers keys must use same protocol"
+        }
+        return ReferenceModeStorageKey(backingKey, containerChildKey)
+    }
+
+    /**
+     * Retrieves [Schema] from the given [Type], if possible.
+     * TODO: declare a common interface.
```
[the remark this answers is not in the document]
Comment (mmandlis) [reply java/arcs/core/storage/CapabilitiesResolverNew.kt]: deleted the todo instead

File: java/arcs/core/storage/CapabilitiesResolverNew.kt [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+            is SingletonType<*> -> if (type.containedType is EntityType) {
+                return (type.containedType as EntityType).entitySchema
+            }
+            is CollectionType<*> -> if (type.collectionType is EntityType) {
+                return (type.collectionType as EntityType).entitySchema
+            }
+            is ReferenceType<*> -> return type.entitySchema!!
+            is EntityType -> return type.entitySchema
+        }
+        throw IllegalArgumentException("Can't retrieve entitySchema of unknown type $type")
```
[the remark this answers is not in the document]
Comment (mmandlis) [reply java/arcs/core/storage/CapabilitiesResolverNew.kt]: right, done! this was copied from the old capabilities resolver, when i knew kotlin even less than today :(

File: javatests/arcs/core/storage/CapabilitiesResolverNewTest.kt [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+        verifyStorageKey(
+            resolver.createStorageKey(inMemoryWithTtls, entityType, handleId),
+            DatabaseStorageKey.Memory::class.java
+        )
+        assertFailsWith<IllegalStateException> {
+            resolver.createStorageKey(onDisk, entityType, handleId)
+        }
+        assertFailsWith<IllegalStateException> {
+            resolver.createStorageKey(onDiskWithTtl, entityType, handleId)
+        }
```
[the remark this answers is not in the document]
Comment (mmandlis) [reply javatests/arcs/core/storage/CapabilitiesResolverNewTest.kt:192]: there is only one exception being thrown there, so probably not.

File: javatests/arcs/core/storage/CapabilitiesResolverNewTest.kt [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+    fun capabilitiesResolver_createWithCustomFactories() {
+        VolatileStorageKey.registerKeyCreator()
+        val resolver = CapabilitiesResolverNew(
+            Options(ArcId.newForTest("test")),
+            listOf(DatabaseStorageKey.Memory.Factory())
+        )
+        verifyStorageKey(
+            resolver.createStorageKey(unspecified, entityType, handleId),
+            VolatileStorageKey::class.java
+        )
```
[the remark this answers is not in the document]
Comment (mmandlis) [reply javatests/arcs/core/storage/CapabilitiesResolverNewTest.kt]: this is so much better, thank you!

File: javatests/arcs/core/storage/CapabilitiesResolverNewTest.kt [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+    @Test
+    fun capabilitiesResolver_failsNonUniqueRegistration() {
+        VolatileStorageKey.registerKeyCreator()
+        assertFailsWith<IllegalArgumentException> {
+            VolatileStorageKey.registerKeyCreator()
+        }
+    }
+
+    @Test
+    fun capabilitiesResolver_fails() {
```
[the remark this answers is not in the document]
Comment (mmandlis) [reply javatests/arcs/core/storage/CapabilitiesResolverNewTest.kt]: renamed

File: javatests/arcs/core/host/AbstractArcHostTest.kt [hunk 15]
```diff
@@ -55,6 +55,7 @@ open class AbstractArcHostTest {
         RamDiskDriverProvider()
     }
 
+
```
[the remark this answers is not in the document]
Comment (mmandlis) [reply javatests/arcs/core/host/AbstractArcHostTest.kt]: deleted. sorry :(

File: javatests/arcs/core/data/CapabilityNewTest.kt [hunk 16]
```diff
@@ -194,4 +194,70 @@ class CapabilityNewTest {
...
+        assertThat(Persistence.ANY.isEquivalent(Persistence.UNRESTRICTED)).isFalse()
+        assertThat(Persistence.UNRESTRICTED.toRange().isEquivalent(Persistence.UNRESTRICTED)).isTrue()
+        assertThat(Persistence.ON_DISK.toRange().isEquivalent(Persistence.ON_DISK.toRange())).isTrue()
+    }
+
+    @Test
+    fun capabilityrange_persistence_contains() {
+        assertThat(Persistence.ON_DISK.toRange().contains(Persistence.ON_DISK.toRange())).isTrue()
+        assertThat(Range(Persistence.ON_DISK, Persistence.IN_MEMORY).contains(
+                      Persistence.ON_DISK)).isTrue()
```
[the remark this answers is not in the document]
Comment (mmandlis) [reply javatests/arcs/core/data/CapabilityNewTest.kt]: done everywhere in the file

File: java/arcs/core/storage/CapabilitiesResolverNew.kt (hunk 1 above)
Comment (mmandlis) [reply java/arcs/core/storage/CapabilitiesResolverNew.kt]: yes, thank you!

File: java/arcs/core/storage/CapabilitiesResolverNew.kt (hunk 2 above)
Comment (mmandlis) [reply java/arcs/core/storage/CapabilitiesResolverNew.kt]: done

File: java/arcs/core/storage/CapabilitiesResolverNew.kt (hunk 3 above)
Comment (mmandlis) [reply java/arcs/core/storage/CapabilitiesResolverNew.kt]: done

File: java/arcs/core/storage/CapabilitiesResolverNew.kt (hunk 5 above)
Comment (mmandlis) [reply java/arcs/core/storage/CapabilitiesResolverNew.kt]: done

File: java/arcs/core/storage/CapabilitiesResolverNew.kt (hunk 6 above)
Comment (mmandlis) [reply java/arcs/core/storage/CapabilitiesResolverNew.kt]: done

File: java/arcs/core/storage/StorageKeyFactory.kt (hunk 7 above)
Comment (mmandlis) [reply java/arcs/core/storage/StorageKeyFactory.kt]: done

File: java/arcs/core/storage/CapabilitiesResolverNew.kt [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+
+    init {
+        require(factoriesList.distinctBy { it.protocol }.size == factoriesList.size) {
+            "Storage keys protocol must be unique $factoriesList."
+        }
+        factories = factoriesList.associateBy { it.protocol }.toMutableMap()
+        CapabilitiesResolverNew.defaultStorageKeyFactories.forEach { (protocol, factory) ->
+            factories.getOrPut(protocol) { factory }
+        }
+    }
```
Comment (bgogul) [inline java/arcs/core/storage/CapabilitiesResolverNew.kt:43]: Use a constructor? 


```suggestion
class CapabilitiesResolverNew(
    private val options: Options,
    private val selector: FactorySelector = SimpleCapabilitiesSelector(),
    private val factories: Map<String, StorageKeyFactory>
) {

    constructor(
         options: Options, 
         selector: FactorySelector = SimpleCapabilitiesSelector(),
         factoriesList: List<StorageKeyFactory> = listOf()
    ) : this(options, selector, CapabilitiesResolverNew.getFactorMap(factoriesList)) 
    
    // This should go into the companion object
    fun getFactorMap(factoriesList: List<StorageKeyFactory>): Map<String, StorageKeyFactory> {
        require(factoriesList.distinctBy { it.protocol }.size == factoriesList.size) {
            "Storage keys protocol must be unique $factoriesList."
        }
        val factories = factoriesList.associateBy { it.protocol }.toMutableMap()
        CapabilitiesResolverNew.defaultStorageKeyFactories.forEach { (protocol, factory) ->
            factories.getOrPut(protocol) { factory }
        }
        return factories
    }
```

File: java/arcs/core/storage/CapabilitiesResolverNew.kt [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+    fun createStorageKey(capabilities: CapabilitiesNew, type: Type, handleId: String): StorageKey {
+        val selectedFactories =
+            factories.filterValues { it.supports(capabilities) }
+
+        if (selectedFactories.isEmpty()) {
+            throw IllegalStateException("Cannot create storage key for handle '$handleId' with " +
+                "capabilities $capabilities")
+        }
+        val factory = selector.select(selectedFactories.values)
+        return createStorageKeyWithFactory(requireNotNull(factory), type, handleId)
```
[the remark this answers is not in the document]
Comment (mmandlis) [reply java/arcs/core/storage/CapabilitiesResolverNew.kt]: good catch, removed.

File: java/arcs/core/storage/CapabilitiesResolverNew.kt [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    fun createStorageKeyWithFactory(
```
[the remark this answers is not in the document]
Comment (mmandlis) [reply java/arcs/core/storage/CapabilitiesResolverNew.kt]: done

File: java/arcs/core/data/CapabilitiesNew.kt [hunk 20]
```diff
@@ -42,15 +41,29 @@ class CapabilitiesNew(capabilities: List<CapabilityNew> = emptyList()) {
...
+    val isEmpty = ranges.isEmpty()
+
+    fun contains(capability: CapabilityNew): Boolean {
+        val otherTag = when (capability.tag) {
+            CapabilityNew.Range.TAG -> (capability as CapabilityNew.Range).min.tag
+            else -> capability.tag
+        }
+        return ranges.find { it.min.tag == otherTag }?.let {
+            it.contains(capability)
+        } ?: false
```
[the remark this answers is not in the document]
Comment (mmandlis) [reply java/arcs/core/data/CapabilitiesNew.kt]: thanks! done

File: java/arcs/core/data/CapabilitiesNew.kt [hunk 21]
```diff
@@ -42,15 +41,29 @@ class CapabilitiesNew(capabilities: List<CapabilityNew> = emptyList()) {
     val isShareable: Boolean?
         get() = getCapability<CapabilityNew.Shareable>()?.let { it.value }
 
+    val isEmpty = ranges.isEmpty()
+
+    fun contains(capability: CapabilityNew): Boolean {
```
[the remark this answers is not in the document]
Comment (mmandlis) [reply java/arcs/core/data/CapabilitiesNew.kt:51]: done

File: java/arcs/core/storage/CapabilitiesResolverNew.kt [hunk 22]
```diff
@@ -0,0 +1,141 @@
...
+    data class Options(val arcId: ArcId)
+
+    fun createStorageKey(capabilities: CapabilitiesNew, type: Type, handleId: String): StorageKey {
+        val selectedFactories =
+            factories.filterValues { it.supports(capabilities) }
+
+        if (selectedFactories.isEmpty()) {
+            throw IllegalStateException("Cannot create storage key for handle '$handleId' with " +
+                "capabilities $capabilities")
+        }
```
[the remark this answers is not in the document]
Comment (mmandlis) [reply java/arcs/core/storage/CapabilitiesResolverNew.kt]: done

File: java/arcs/core/storage/CapabilitiesResolverNew.kt (hunk 17 above)
Comment (mmandlis) [reply java/arcs/core/storage/CapabilitiesResolverNew.kt]: done.
one reason i can think of why this is preferable (despite being more verbose) it that the class declaration clearly shows all data members. are there other reasons?

**Pull request merged**